### ISSUE TITLE
Oauth PKCE flow - code_verifier & code_challenge

### ISSUE THREAD

**Issue by JensDebergh** [omniauth/omniauth#1001]
Hey there!

I'm having some troubles implementing an oauth provider using PKCE flow for oauth.

I'm at the point where I receive the `authorization_code`  back from the authorization server and I can exchange it for an `access_token` & `refresh_token`. But to get this authentication code I needed to generate a `code_challenge`.

As I understand it the code challenge is generate by generating a `code_verifier`

```
code_verifier = SecureRandom.hex(24)
code_challenge = Base64.encode64(Digest::SHA256.digest(code_verifier))
```
This `code_challenge` is associated on the authorization server with the `authentication_code` they generated.

To get an access token I have to send in the `authorization_code` + `code_verifier` so they can validate on their end.

The problem is I can only validate in the callback phase which generates a new code_verifier and it never matches the one from the request phase.

Am I doing something wrong? Does omniauth handle this for me?

Any push in the right direction would be greatly appreciated.

I can blog about this issue afterwards

Kind regards
Jens

Documentation:

https://documentation.ibanity.com/ponto-connect/products#access-authorization

```
# frozen_string_literal: true

require 'omniauth/strategies/oauth2'

module OmniAuth
  module Strategies
    class Ponto < OmniAuth::Strategies::OAuth2
      option :name, :ponto

      option :client_options, authorize_url: 'https://sandbox-authorization.myponto.com/oauth2/auth'

      option :authorize_params,
             response_type: 'code',
             scope: 'ai pi offline_access',
             state: SecureRandom.hex(24)


      def request_phase
        redirect client.auth_code.authorize_url(authorize_params)
      end

      def callback_phase
        authorization_code = request.params['code']

        token = Ibanity::PontoConnect::Token.create(
          authorization_code: authorization_code,
          redirect_uri: redirect_uri,
          code_verifier: code_verifier
        )
      end

      uid { user_data['id'] }

      extra do
        {
          'name' => user_data['name']
        }
      end

      def authorize_params
        super
          .merge(
            code_challenge: code_challenge,
            code_challenge_method: 'S256',
            redirect_uri: redirect_uri
          )
      end

      def code_verifier
        @code_verifier ||= SecureRandom.hex(24)
      end

      def code_challenge
        @code_challenge ||= Base64.encode64(Digest::SHA256.digest(code_verifier))
      end

      def redirect_uri
        'https://7436f4c8.ngrok.io/auth/ponto/callback' || callback_url
      end
    end
  end
end

```

**Comment by tkawa** [comment]
How about storing `code_verifier` value on `session` in `request_phase`?

**Comment by pierre-pretorius** [comment]
@JensDebergh Did you manage to get PKCE working?

**Comment by BobbyMcWho** [comment]
Hello, we just released [omniauth-oauth2 1.7.0](https://github.com/omniauth/omniauth-oauth2/releases/tag/v1.7.0) yesterday which will enable pkce to be used in your strategy.
